### ISSUE TITLE
Docker chevdor/substrate-api-sidecar is not working. 'Connection dropped by remote peer.'

### ISSUE THREAD

**Issue by schadokar** [paritytech/substrate-api-sidecar#318]
**Description**

I am using the `chevdor/substrate-api-sidecar:latest` docker image.  
It was working perfectly fine but from last couple of days it is unable to connect to the public node.  
It used to connect to `ws://rpc.polkadot.io`. But now logs show `wss://rpc.polkadot.io`.  

I changed it to `ws` by configuring environment variable `SAS_SUBSTRATE_WS_URL:  ws://rpc.polkadot.io` but it is still not working. By default it is always connecting to `wss`. 

### Errors
When trying to connect `ws://rpc.polkadot.io`.  
```
Attaching to polkadot-sidecar
polkadot-sidecar | SAS:
polkadot-sidecar |   📦 EXPRESS:
polkadot-sidecar |      ✅ LOG_MODE: errors
polkadot-sidecar |      ✅ BIND_HOST: 0.0.0.0
polkadot-sidecar |      ✅ PORT: 8080
polkadot-sidecar |   📦 SUBSTRATE:
polkadot-sidecar |      ✅ NAME: Parity Polkadot node
polkadot-sidecar |      ✅ WS_URL: ws://rpc.polkadot.io
polkadot-sidecar | Connecting to Parity Polkadot node at ws://rpc.polkadot.io
polkadot-sidecar | 2020-10-14 14:30:33          API-WS: disconnected from ws://rpc.polkadot.io code: '1006' reason: 'connection failed'
polkadot-sidecar | _Event {
polkadot-sidecar |   type: 'error',
polkadot-sidecar |   isTrusted: false,
```

When trying to connect `wss://rpc.polkadot.io`.  
```
Creating network "nodepolkadot_default" with the default driver
Creating polkadot-sidecar ...
Creating polkadot-sidecar ... done
Attaching to polkadot-sidecar
polkadot-sidecar | SAS:
polkadot-sidecar |   📦 EXPRESS:
polkadot-sidecar |      ✅ LOG_MODE: errors
polkadot-sidecar |      ✅ BIND_HOST: 0.0.0.0
polkadot-sidecar |      ✅ PORT: 8080
polkadot-sidecar |   📦 SUBSTRATE:
polkadot-sidecar |      ✅ NAME: Parity Polkadot node
polkadot-sidecar |      ✅ WS_URL: wss://rpc.polkadot.io
polkadot-sidecar | Connecting to Parity Polkadot node at wss://rpc.polkadot.io
polkadot-sidecar | 2020-10-14 14:56:54        API/INIT: RPC methods not decorated: grandpa_proveFinality, grandpa_subscribeJustifications, grandpa_unsubscribeJustifications
polkadot-sidecar | 2020-10-14 14:56:54        RPC-CORE: getMetadata(at?: BlockHash): Metadata:: createType(Metadata):: Unable to create Enum via index 12, in V0, V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11
polkadot-sidecar | 2020-10-14 14:56:54        API/INIT: Error: FATAL: Unable to initialize the API: createType(Metadata):: Unable to create Enum via index 12, in V0, V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11
polkadot-sidecar |     at EventEmitter.value (/usr/src/app/node_modules/@polkadot/api/base/Init.js:77:25)
polkadot-sidecar |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
polkadot-sidecar | Error: FATAL: Unable to initialize the API: createType(Metadata):: Unable to create Enum via index 12, in V0, V1, V2, V3, V4, V5, V6, V7, V8, V9, V10, V11
polkadot-sidecar |     at EventEmitter.value (/usr/src/app/node_modules/@polkadot/api/base/Init.js:77:25)
polkadot-sidecar |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
polkadot-sidecar | 2020-10-14 14:57:54          API-WS: disconnected from wss://rpc.polkadot.io code: '1006' reason: 'Connection dropped by remote peer.'
```

**Steps to Reproduce**
`docker-compose.yml`

```
version: "2"
services:
  polkadot:
    image: chevdor/substrate-api-sidecar:latest
    restart: on-failure
    ports:
      - "8080:8080"
    container_name: polkadot-sidecar
    environment:
      SAS_SUBSTRATE_WS_URL: ws://rpc.polkadot.io
```

I tried all the solution of [issue304](https://github.com/paritytech/substrate-api-sidecar/issues/304#issuecomment-706522586) but it is not working for me.

Thanks

**Comment by emostov** [comment]
Hi - thank you fo the detailed issue.

If I understand correctly, It looks like the @chevdor docker image has not been updated in 4 months. You will need to find a more up to date image (preferably v1.0.0) and [build your own image](https://github.com/paritytech/substrate-api-sidecar#docker). I am not sure if that will fix your exact issue, but either way your version of Sidecar is out of date and will not work.

Sidecar needs upgrades often since new types in runtime upgrades require that any API users update their type definitions.

**Comment by chevdor** [comment]
Yes, the docker image I made was more of a one-shot test to support the PR at the time.
I do not build the image automatically or frequently so the image is almost guaranteed to be outdate if working at all.

# Background on the issue
The substrate-api-sidecar project is depending on the PolkadotJS API. This API is closely following the frequent evolutions of substrate/polkadot. That means that all projects need to update their version dependencies in order to use the latest API version. The old image you test is using a new ancient version of the API that no longer works with recent nodes. This is why you see the errors you see. 

# New image
That being said, I don't mind a refresh once in a while so I built an updated version and pushed it. That should solve your problem until the next updates for the nodes/API.

I am no longer sure that the `default` and following the doc tries to connect to your localhost, that will not work unless you run a local node.

Here is a test with the new version of the image I pushed a few minutes ago:
```
$ docker run --rm -it -p 8080:8080 -e SAS_SUBSTRATE_WS_URL=wss://rpc.polkadot.io chevdor/substrate-api-sidecar:1.0.0
SAS:
  📦 LOG:
     ✅ LEVEL: "info"
     ✅ JSON: false
     ✅ FILTER_RPC: false
     ✅ STRIP_ANSI: false
  📦 SUBSTRATE:
     ✅ WS_URL: "wss://rpc.polkadot.io"
  📦 EXPRESS:
     ✅ BIND_HOST: "0.0.0.0"
     ✅ PORT: 8080
2020-10-29 09:13:09 info: Connected to chain Polkadot on the parity-polkadot client at wss://rpc.polkadot.io
2020-10-29 09:13:09 info: wss://rpc.polkadot.io is a public node. Too many users will overload this public endpoint. Switch to a privately hosted node when possible.
2020-10-29 09:13:09 info: Listening on http://0.0.0.0:8080/
```

# Build your own
In general however, you should not count on the image I built unless in the unlikely case where you test locally with the polkadot/substrate/api version that was the latest at the time of publishing the image. Nonetheless, thank you the heads up on the issue.

# Conclusion
In conclusion, I think we can close this issue as the new image has been published and works.
And make you took note that the image will become outdated again in the future so building your own is both very simple (and documented) and the safest option you have to ensure you are using the very latest version of the substrate-api-sidecar.

@emostov I noticed that the Dockerfile needs a little refresher. I will open a new PR for that.

**Comment by schadokar** [comment]
Thanks @emostov and @chevdor . I build the image from the repo this time and it is working perfectly fine.
